libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -10,157 +10,269 @@
             {
-                var entityKey = GetEntityKey(key);
-                var result = await this.Table.ExecuteAsync(TableOperation.Retrieve<StoreItemEntity>(entityKey.PartitionKey, entityKey.RowKey)).ConfigureAwait(false);
-                if ((HttpStatusCode)result.HttpStatusCode == HttpStatusCode.OK)
+                var ek = new EntityKey(key);
+                var tableEntity = await table.ExecuteAsync(TableOperation.Retrieve<DynamicTableEntity>(ek.PartitionKey, ek.RowKey)).ConfigureAwait(false);
+
+                if (tableEntity.HttpStatusCode == (int)HttpStatusCode.OK)
                 {
-                    var value = ((StoreItemEntity)result.Result).AsObject();
-                    var valueStoreItem = value as IStoreItem;
-                    if (valueStoreItem != null)
-                    {
-                        valueStoreItem.eTag = result.Etag;
-                    }
-                    storeItems.Add(new KeyValuePair<string, object>(key, value));
+                    // re-create expected object
+                    StorageEnvelope envelope = StorageEnvelope.AsStoreItemEntity(tableEntity);
+                    return new KeyValuePair<string, object>(key, envelope.StoreItem);
                 }
-            }
-            return storeItems;
-        }
 
+                return new KeyValuePair<string, object>();
+            });
 
+            return (await Task.WhenAll(readTasks).ConfigureAwait(false))
+                .Where(kv => kv.Key != null);
+        }
+
+        /// <summary>
+        /// Saves store items to storage.
+        /// </summary>
+        /// <param name="changes">Map of items to write to storage.</param>
+        /// <returns></returns>
         public async Task Write(IEnumerable<KeyValuePair<string, object>> changes)
         {
-            foreach (var change in changes)
+            if (changes == null) throw new ArgumentNullException(nameof(changes));
+
+            var storeItems = changes.Select(kv => new StorageEnvelope(new EntityKey(kv.Key), kv.Value));
+            var bogusEtagKeys = storeItems.Where(item => item.ETag != null && item.ETag.Length == 0);
+            if (bogusEtagKeys.Any())
             {
-                var entityKey = GetEntityKey(change.Key);
-                var newValue = change.Value;
-                StoreItemEntity entity = new StoreItemEntity(entityKey, newValue);
-                if (entity.ETag == null || entity.ETag == "*")
-                {
-                    var result = await this.Table.ExecuteAsync(TableOperation.InsertOrReplace(entity)).ConfigureAwait(false);
-                }
-                else if (entity.ETag.Length > 0)
-                {
-                    var result = await this.Table.ExecuteAsync(TableOperation.Replace(entity)).ConfigureAwait(false);
-                }
-                else
+                throw new ArgumentException("Bogus etag in items with key: " + string.Join(", ", bogusEtagKeys.Select(o => o.Key)));
```

Comment:
Would just clean this exception message up a bit. Maybe something like?

`"Invalid ETag values detected for items with the following keys:"`